### ISSUE TITLE
Move output base somewhere under the user's home directory

### ISSUE THREAD

**Issue by jsharpe** [bazelbuild/bazel#11100]
### Description of the problem / feature request:

The permissions on the directory created by the sandbox_base option are world readable by default. When this is set to /dev/shm on a shared user system this exposes the build files to all users during the build.

### Feature requests: what underlying problem are you trying to solve with this feature?

Remove world readable permissions from the directory created by sandbox_base

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

bazel build //... --sandbox_base=/dev/shm
Observe permissions on sandbox directory in /dev/shm are world readable

### What operating system are you running Bazel on?

centos 7

**Comment by jmmv** [comment]
From https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java;l=331;drc=eae93f643ccfe19df24bb1dd837c45056a15c274 , which is the implementation that should be used in your machine:

We do create directories with 0777 permissions... but that's the right thing to do for any code that wants to honor the user's umask.

So let's start there... what's your umask? Is it correctly set?

**Comment by jsharpe** [comment]
My umask is 0022 so that explains the behaviour I am seeing.

Having said that you state that creating directories with 0777 permissions is the correct thing to do for any code that wants to honor umask but is this the correct assumption to make for all instances of directory creation in bazel? 

Given that bazel should work for any users reasonable setting of umask (0000, 0002, 0007, 0022, 0027, 0077?) then I'd suggest that any operations that a hermetic build rule can only rely on being able to use the user permissions themselves, as umask may have masked out access for group or other, and so the only reason for respecting any broader permission set is for external tooling to be able to access them outside of bazel. As the sandbox is purely an internal implementation detail of bazel's build isolation then surely the directories created for the purposes of sandbox_base should only be accessible by the user running bazel or am I missing something more subtle here?

**Comment by jmmv** [comment]
I think that makes sense (making the sandbox tree explicitly `0700` or something like that). But what about the many other directories that Bazel creates?

For example: should we do the same for the output base? That directory is created under `/var/tmp`, and as far as I can see, we use `0755` for that: https://source.bazel.build/bazel/+/master:src/main/cpp/blaze_util_posix.cc;l=489;drc=759170fc56c34356a46fb35ab1f108c8535d861f. So... even if we protected the sandbox base, we would still expose the output tree itself. Which means this issue may be more problematic than originally stated because it affects default behavior, not just when you reconfigure the sandbox base location, right?

**Comment by jmmv** [comment]
Another issue that surfaced in #7435: the sandbox allows blanket write access to `tmp` paths, so by putting the output base or the sandbox base under `tmp`, we essentially disable some strictness that the sandbox is trying to enforce: i.e. separate actions could write to the execroot of other actions.

I think we really should move the output base outside of tmp and put it beneath the user's home directory.

**Comment by jmmv** [comment]
#8407 describes the same problem I mentioned about the sandbox. I'm making this bug more general to track all related issues. @philwo was proposing ~/Library/caches/bazel or something like that, which I think would make sense.

**Comment by jablin** [comment]
I'd like you to consider that home directories might be on NFS.

I really don't know how to configure Bazel in this case. What's the recommended solution?

**Comment by philwo** [comment]
@jablin If e.g. `/scratch` is a fast local disk, you can use `bazel --output_user_root /scratch/bazel build [...]` or even put it into your `~/.bazelrc` so that it always takes effect:

```
echo "startup --output_user_root /scratch/bazel" >> ~/.bazelrc
```

This should cause Bazel to put all files that it generates underneath that directory.

**Comment by jablin** [comment]
Thanks, @philwo , for your suggestion. I've already tried that. It works. For a single uses, that is.

However, it does _not_ work for multiple users per workstation.

Maybe enhance it like this?

    echo "startup --output_user_root /scratch/bazel-$LOGNAME" >> ~/.bazelrc

Gotchas:
* need for individual configuration
* There will be lots of `/scratch/bazel-*/` directories. With lots of redundant files, I guess.

**Comment by philwo** [comment]
@meisterT Do you know a good way to solve this? 🤔

**Comment by meisterT** [comment]
not really, cc @larsrc-google

**Comment by larsrc-google** [comment]
The number of redundant files would be no different from when they're in user's home dirs. A slightly better way to configure  @philwo's suggestion would be

```
echo "startup --output_user_root /scratch/bazel-$USER" >> ~/.bazelrc
```
There's no expansion in the bazelrc's (which is probably a good state for them), so this can't easily be globally configured.

There's also no way to use `--output_user_root` to get the default behaviour of `<tmpdir>/_bazel_$USER` in a blazerc, which is not great.

**Comment by jablin** [comment]
> The number of redundant files would be no different from when they're in user's home dirs.

What I meant to say: If more than one user compiles in the same workspace:
* each user has to compile; user's won't share result (except when using --disk_cache or --remote_cache); the bazel-* symlinks will be switched back and forth
* the build results of each user will be identical but stored separately (i.e. many times, therefor in a redundant way) because each user uses a different output workspace

**Comment by larsrc-google** [comment]
Can you describe the setup where multiple users compile in the same workspace? It sounds like an uncommon setup.

**Comment by jablin** [comment]
> Can you describe the setup where multiple users compile in the same workspace?

Well, the developers just do it. `umask 02` in your dot files and you're done. Of course, at a given point in time, only one developers works on it. They _don't_ work on a single workspace in parallel.

**Comment by larsrc-google** [comment]
Ok. It sounds like you just want everybody to use the same workspace, but it doesn't work? What actually happens when you follow philwo's suggestion above?
